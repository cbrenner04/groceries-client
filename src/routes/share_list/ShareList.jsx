import React from 'react';
import Async from 'react-async';
import PropTypes from 'prop-types';

import ShareListForm from './containers/ShareListForm';
import { fetchData } from './utils';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

function ShareList(props) {
  return (
    <Async promiseFn={fetchData} listId={props.match.params.list_id} history={props.history}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {data => (
          <ShareListForm
            history={props.history}
            name={data.name}
            invitableUsers={data.invitableUsers}
            listId={data.listId}
            userIsOwner={data.userIsOwner}
            pending={data.pending}
            accepted={data.accepted}
            refused={data.refused}
            userId={data.userId}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
}

ShareList.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      list_id: PropTypes.string,
    }),
  }),
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
};

export default ShareList;
