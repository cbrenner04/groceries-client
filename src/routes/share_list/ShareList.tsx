import React, { type ReactNode } from 'react';
import Async, { type PromiseFn } from 'react-async';
import { useNavigate, useParams } from 'react-router-dom';

import Loading from 'components/Loading';
import type { IListUser, IUsersList } from 'typings';

import ShareListForm from './containers/ShareListForm';
import { fetchData } from './utils';
import UnknownError from '../error_pages/UnknownError';

const ShareList: React.FC = (): React.JSX.Element => {
  const navigate = useNavigate();
  const { list_id: pListId } = useParams();

  return (
    // TODO: figure out typings for PromiseFn
    <Async promiseFn={fetchData as unknown as PromiseFn<void>} listId={pListId} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: {
          name: string;
          invitableUsers: IListUser[];
          listId: string;
          userIsOwner: boolean;
          pending: IUsersList[];
          accepted: IUsersList[];
          refused: IUsersList[];
          userId: string;
        }): ReactNode => (
          <ShareListForm
            name={data.name}
            invitableUsers={data.invitableUsers}
            listId={data.listId}
            userIsOwner={data.userIsOwner}
            pending={data.pending}
            accepted={data.accepted}
            refused={data.refused}
            userId={data.userId}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default ShareList;
