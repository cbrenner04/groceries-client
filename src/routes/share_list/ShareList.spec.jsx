import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import ShareList from './ShareList';
import axios from '../../utils/api';

describe('ShareList', () => {
  const renderShareList = () => {
    const history = createMemoryHistory();
    const props = {
      history,
      match: {
        params: {
          list_id: '1',
        },
      },
    };
    return render(
      <Router history={history}>
        <ShareList {...props} />
      </Router>,
    );
  };

  it('renders loading when data fetch is not complete', () => {
    const { container, getByText } = renderShareList();

    expect(container).toMatchSnapshot();
    expect(getByText('Loading...')).toBeVisible();
  });

  it('renders unknown error component when error occurs', async () => {
    axios.get = jest.fn().mockRejectedValue({ response: { status: 400 } });
    const { container, getByRole } = renderShareList();
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
    expect(getByRole('button')).toHaveTextContent('refresh the page');
  });

  it('renders ShareList when data fetch is successful', async () => {
    axios.get = jest.fn().mockResolvedValue({
      data: {
        accepted: [
          { user: { id: 'id1', email: 'foo@example.com' }, users_list: { id: 'id1', permissions: 'read' } },
          { user: { id: 'id4', email: 'foobaz@example.com' }, users_list: { id: 'id4', permissions: 'write' } },
        ],
        pending: [{ user: { id: 'id2', email: 'bar@example.com' }, users_list: { id: 'id2', permissions: 'write' } }],
        refused: [{ user: { id: 'id3', email: 'baz@example.com' }, users_list: { id: 'id3', permissions: 'read' } }],
        current_user_id: 'id4',
        user_is_owner: true,
        invitable_users: [{ id: 'id5', email: 'foobar@example.com' }],
        list: {
          name: 'foo',
          id: 'id1',
        },
      },
    });
    const { container, getByTestId, queryByTestId } = renderShareList();
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
    expect(getByTestId('invite-user-id5')).toHaveTextContent('foobar@example.com');
    expect(getByTestId('accepted-user-id1')).toHaveTextContent('foo@example.com');
    expect(queryByTestId('accepted-user-id4')).toBeNull();
    expect(getByTestId('pending-user-id2')).toHaveTextContent('bar@example.com');
    expect(getByTestId('refused-user-id3')).toHaveTextContent('baz@example.com');
  });
});
