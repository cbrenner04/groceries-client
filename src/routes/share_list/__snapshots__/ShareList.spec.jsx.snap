// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareList renders ShareList when data fetch is successful 1`] = `
<div>
  <div>
    <a
      class="float-end"
      href="/lists"
    >
      Back to lists
    </a>
    <h1>
      Share 
      foo
    </h1>
    <form
      class="pt-3 pb-3"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="new-email"
        >
          Enter an email to invite someone to share this list:
        </label>
        <input
          class="form-control"
          id="new-email"
          name="new-email"
          placeholder="jane.smith@example.com"
          type="email"
          value=""
        />
      </div>
      <div
        class="d-grid gap-2 mt-3"
      >
        <button
          class="btn btn-success"
          type="submit"
        >
          Share List
        </button>
      </div>
    </form>
    <p
      class="text-lead"
    >
      Or select someone you've previously shared with:
    </p>
    <div
      class="mb-5 list-group"
    >
      <div
        data-test-id="invite-user-id5"
      >
        <button
          class="btn btn-link list-group-item list-group-item-action"
        >
          foobar@example.com
        </button>
      </div>
    </div>
    <div
      class="m-0 row"
    >
      <h2
        class="text-capitalize pe-0 w-auto"
      >
        pending
      </h2>
      <i
        class="far fa-question-circle float-end text-secondary ps-0 w-auto"
        data-test-id="pending-popover"
      />
    </div>
    <div
      class="mb-4 list-group"
    >
      <div
        class="list-group-item"
        data-test-id="pending-user-id2"
        style="display: block;"
      >
        <div
          class="row"
        >
          <div
            class="pt-1 col-md-6"
          >
            bar@example.com
          </div>
          <div
            class="pt-1 col-md-4"
          >
            <span
              class="badge"
              data-test-id="perm-write"
              variant="success"
            >
              write
            </span>
          </div>
          <div
            class="col-md-2"
          >
            <div
              class="float-end btn-group"
              role="group"
            >
              <button
                class="p-0 me-4 btn btn-link"
                data-test-id="toggle-permissions"
                type="button"
              >
                <i
                  class="fas fa-angle-double-down fa-2x text-warning"
                />
              </button>
              <button
                class="p-0 btn btn-link"
                data-test-id="remove-share"
                type="button"
              >
                <i
                  class="fa fa-trash fa-2x text-danger"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="m-0 row"
    >
      <h2
        class="text-capitalize pe-0 w-auto"
      >
        accepted
      </h2>
      <i
        class="far fa-question-circle float-end text-secondary ps-0 w-auto"
        data-test-id="accepted-popover"
      />
    </div>
    <div
      class="mb-4 list-group"
    >
      <div
        class="list-group-item"
        data-test-id="accepted-user-id1"
        style="display: block;"
      >
        <div
          class="row"
        >
          <div
            class="pt-1 col-md-6"
          >
            foo@example.com
          </div>
          <div
            class="pt-1 col-md-4"
          >
            <span
              class="badge"
              data-test-id="perm-read"
              variant="primary"
            >
              read
            </span>
          </div>
          <div
            class="col-md-2"
          >
            <div
              class="float-end btn-group"
              role="group"
            >
              <button
                class="p-0 me-4 btn btn-link"
                data-test-id="toggle-permissions"
                type="button"
              >
                <i
                  class="fas fa-angle-double-up fa-2x text-warning"
                />
              </button>
              <button
                class="p-0 btn btn-link"
                data-test-id="remove-share"
                type="button"
              >
                <i
                  class="fa fa-trash fa-2x text-danger"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      
    </div>
    <h2>
      Refused
    </h2>
    <div
      class="list-group"
    >
      <div
        class="list-group-item"
        data-test-id="refused-user-id3"
        style="display: block;"
      >
        <div
          class="row"
        >
          <div
            class="pt-1 col-md-6"
          >
            baz@example.com
          </div>
          <div
            class="pt-1 col-md-4"
          />
          <div
            class="col-md-2"
          >
            <div
              class="float-end btn-group"
              role="group"
            >
              <button
                class="p-0 me-3 btn btn-link"
                data-test-id="refresh-share"
                type="button"
              >
                <i
                  class="fa fa-redo fa-2x text-primary"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ShareList renders loading when data fetch is not complete 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center;"
  >
    <div
      class="spinner-border"
      role="status"
      style="width: 100px; height: 100px; position: absolute; top: 50%;"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`ShareList renders unknown error component when error occurs 1`] = `
<div>
  <h1>
    Something went wrong!
  </h1>
  <h2>
    We are currently unable to render this page.
  </h2>
  <p>
    Please check your connection and try to Â 
    <button
      style="background-color: transparent; cursor: pointer; display: inline; margin: 0px; padding: 0px; color: rgb(0, 123, 255); text-decoration: none;"
    >
      refresh the page
    </button>
  </p>
</div>
`;
